topics/data-science/tutorials/python-functions/tutorial.md

```diff
@@ -0,0 +1,300 @@
+
+### Documentation
+
+Documenting your code is extremely, *extremely*, **extremely** important to do. We all forget what we're doing, it's only normal, so documenting what you're doing is key to being able to restart work later.
+
+```python
+def fahr_to_kelvin(temp_f):
+    """
+    Converts a temperature from Fahrenheit to Kelvin
+
+    temp_f: the temperature in Fahrenheit
+
+    returns the temperature in Celsius
+    """
+    temp_c = fahr_to_celsius(temp_f)
+    temp_k = celsius_to_kelvin(temp_c)
+    return temp_k
+
+print('boiling point of water in Kelvin:', fahr_to_kelvin(212.0))
+```
+
+For a function this small, with such a descriptive name (`fahr_to_kelvin`) it feels quite obvious what the function should do, what inputs it takes, what outputs it produces. However
+
+> ### {% icon question %} Question: Converting statements to functions
+> A lot of what you'll do in programing is to turn a procedure that you want to do, into statements and a function.
+>
+> Fill in the missing portions of this function, two average numbers. Then use it to find the average of 32326 and 631
+>
+> > ### {% icon solution %} Solution
+> > ```
+> > def average2(a, b):
+> >     c = (a + b) / 2
+> >     return c
+> > ```
+> {: .solution}
+{: .question}
+
+```
+# Test out solutions here!
+def average2(a, b):
+    c =
+    return c
+
+print(average2(32326, 631))
+```
+
+> ### {% icon question %} Question: A more complicated example
+>
+> The formula for a 90° triangle can be expressed as: $c = \sqrt{a^2 + b^2}$
+>
+> 1. Write a function which takes `a` and `b`, and calculates `c`
+> 2. Name this function "pythagorus"
+> 3. Remember to import math, if you haven't already
+>
+> > ### {% icon solution %} Solution
+> > ```
+> > def pythagorus(a, b):
+> >     c = math.sqrt(math.pow(a, 2) + math.pow(b, 2))
+> >     return c
+> > ```
+> {: .solution}
+{: .question}
+
+```
+# Test out solutions here!
+
+
+print(pythagorus(1234, 4321)) # Should return 4493.750883170984
+```
+
+
+### Variable Scope
+
+In composing our temperature conversion functions, we created variables inside of those functions, `temp`, `temp_c`, `temp_f`, and `temp_k`. We refer to these variables as local variables because they no longer exist once the function is done executing. If we try to access their values outside of the function, we will encounter an error:
+
+```python
+print('Again, temperature in Kelvin was:', temp_k)
+```
+
+If you want to reuse the temperature in Kelvin after you have calculated it with fahr_to_kelvin, you can store the result of the function call in a variable:
+
+```python
+temp_kelvin = fahr_to_kelvin(212.0)
+print('temperature in Kelvin was:', temp_kelvin)
+```
+
+Watch out for scope issues:
+
+- Variables created inside a function will stay inside the function
+- Variables created outside of the function can be accessible inside the function, but you should not do this!
```

Comment:
maybe add:

- it is bad practice and will later cause errors when working with modules or testing big projects.